### PR TITLE
Initial draft GitBook.

### LINKED ISSUE #505: Clear documentation for machine and cluster actuator interface
The machine actuator interface is not defined clearly here: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machine/actuator.go#L25 It's confusing to the implementer of the interface about the expected behavior. The same applies to cluster actuator. 

For instance: 
- Existed
  - What is existing? VM is created, or started, or network stack is ready, or the service is started and healthy? 
- Create and Update
  - it looks like both Update and Create are idempotent, then what is the difference between them. https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machine/controller.go#L137 and https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machine/controller.go#L148  

Also, is it a good idea to put the idempotent burden to actuator implementer? especially for Create. Create does the whole lot of thing, and doing it idempotent isn't a trivial job. If the VM doesn't exist, it's fine. However, if the target VM exists, what should we do? Those logics are very common to all different actuators. Should we handle this in in controller? Actuator should just become a stateless cloud provider abstraction.

### PR DESCRIPTION
**Pull request by davidewatson**
**What this PR does / why we need it**:

The Cluster API has seen a fair number of changes over the course of the past year. At the same time, there has been a significant increase in the number of providers. Clear developer documentation is needed in order to make implementation easier and maintain consistent behavior.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Begins to fix #505 

**Special notes for your reviewer**:

I am going to push the generated code in a second commit for ease of review. It may also make sense for me to break this up further to both obtain consensus for this PR, and facilitate further writing.

To view this GitBook locally while we sort out which account will be used to serve the final product, you may [install GitBook](https://toolchain.gitbook.com/setup.html) and then:

```
cd $GOPATH/src/sigs.k8s.io/cluster-api
git fetch origin pull/566/head:gitbook
git checkout gitbook
cd docs/book
gitbook install
gitbook build
gitbook serve
```

You may then view the book locally `open http://localhost:4000`...

**Release note**:
```release-note
Initial developer documentation.
```

### CODE REVIEW

File: docs/book/common_code/machineset_controller.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+}
+```
+{% endmethod %}
+
+{% method %}
+## MachineSetTemplateSpec
+
+{% sample lang="go" %}
+```go
+// MachineTemplateSpec describes the data a machine should have when created from a template
```
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/common_code/machineset_controller.md]: 👍 Since this is code and not strictly documentation I think I should create separate PR for these changes?

File: docs/book/common_code/machineset_controller.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+```go
+// MachineSetSpec defines the desired state of MachineSet
+type MachineSetSpec struct {
+	// Replicas is the number of desired replicas.
+	// This is a pointer to distinguish between explicit zero and unspecified.
+	// Defaults to 1.
+	// +optional
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// Minimum number of seconds for which a newly created machine should be ready.
```
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/common_code/machineset_controller.md]: 👍 Since this is code and not strictly documentation I think I should create separate PR for these changes?

File: docs/book/common_code/machine_controller.md [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+{% method %}
+## MachineSpec
+
+The `ProviderSpec` is recommended to be a serialized API object in a format 
+owned by that provider. This will allow the configuration to be strongly typed,
+versioned, and have as much nested depth as appropriate. These provider-specific
+API definitions are meant to live outside of the Machine API, which will allow
+them to evolve independently of it. Attributes like instance type, which 
+network to use, and the OS image all belong in the `ProviderSpec`.
```
Comment (detiber) [inline docs/book/common_code/machine_controller.md]: Should `ProviderSpec` also be called out for the Cluster Controller doc above as well?

Comment (detiber) [reply docs/book/common_code/machine_controller.md]: Would it also be good to call out the `ProviderStatus` in a similar way as well?

File: docs/book/common_code/machine_controller.md [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+	// +optional
+	ConfigSource *corev1.NodeConfigSource `json:"configSource,omitempty"`
+}
+```
+{% endmethod %}
+
+{% method %}
+## MachineStatus
+
+Note that the **NodeRef** field may not exist, in particular in the
```
Comment (detiber) [inline docs/book/common_code/machine_controller.md]: I'd recommend against saying 'may not exist' and instead say that 'it may be unused' or 'it may not be set'

File: docs/book/common_code/machine_controller.md [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+case of providers which manage remote Kubernetes clusters.
+
+{% sample lang="go" %}
+```go
+// MachineStatus defines the observed state of Machine
+type MachineStatus struct {
```
Comment (detiber) [inline docs/book/common_code/machine_controller.md]: Is there some tooling to sync these structs with the current versions? If not, we may want to track that as a followup item to avoid drift.

File: docs/book/common_code/machine_controller.md [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+  actuator `Create()` method.
+
+{% panel style="warning", title="Machines depend on Clusters" %}
+The Machine actuator methods expect both a `Cluster` and a `Machine` to be
+passed in. While there is not a strong link between `Cluster`s and `Machine`s,
+the machine controller will determine which cluster to pass by looking for a 
+`Cluster` in the same namespace as the `Machine`
+
+There are two consequences of this:
+ - There must be at exactly one `Cluster` per namespace.
```
Comment (detiber) [inline docs/book/common_code/machine_controller.md]: I'd suggest rewording this to say "There must no be more than one `Cluster` per namespace". The current wording seems to suggest that namespaces without a `cluster` would also be problematic.

File: docs/book/provider_implementations/building_running_and_testing.md [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+ # Deploy controller in the configured Kubernetes cluster in ~/.kube/config
+ deploy: manifests
+-       kubectl apply -f config/crds
+-       kustomize build config/default | kubectl apply -f -
++       cat provider-components.yaml | kubectl apply -f -
+ 
+ # Generate manifests e.g. CRD, RBAC etc.
+ manifests:
+-       go run vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go all
++       go run vendor/sigs.k8s.io/controller-tools/cmd/controller-gen/main.go crd
```
Comment (detiber) [inline docs/book/provider_implementations/building_running_and_testing.md]: This probably isn't ideal. Ideally we'd be able to leverage more of the auto generation from kubebuilder. For the AWS implementation we disabled most of it for expedience as we were working on break-fix after the initial CRD migration.

File: docs/book/provider_implementations/generate_crds.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+{% sample lang="bash" %}
+```bash
+kubebuilder create api --group solas --version v1alpha1 --kind SolasClusterProviderSpec
+kubebuilder create api --group solas --version v1alpha1 --kind SolasClusterProviderStatus
+```
+```bash
+kubebuilder create api --group solas --version v1alpha1 --kind SolasMachineProviderSpec
+kubebuilder create api --group solas --version v1alpha1 --kind SolasMachineProviderStatus
+```
+
```
Comment (detiber) [inline docs/book/provider_implementations/generate_crds.md]: Should there be additional `kubebuilder create` commands to create the controllers and the manager scaffolding as well?

File: docs/book/common_code/machine_controller.md (hunk 5 above)
Comment (davidewatson) [reply docs/book/common_code/machine_controller.md]: Agreed that duplicating this code is a Bad Idea (tm). I've found two GitBook plugins which support including code snippets: [include-codeblock](https://plugins.gitbook.com/plugin/include-codeblock) and [snippet](https://plugins.gitbook.com/plugin/snippet). Both of them assume Doxygen [markers](https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdsnippet), which seems reasonable to me. If we are okay with adding those to the source I'll take a look.

File: docs/book/provider_implementations/generate_crds.md (hunk 8 above)
Comment (davidewatson) [reply docs/book/provider_implementations/generate_crds.md]: I don't think so. The manager is created automatically, and `kubebuilder create` will prompt the user for whether they would like controllers to be generated as well. Since we reuse the Cluster API controllers, we only need to register the existing controllers with the manager.

Which reminds me: I am not sure if we ever define what a manager is...

File: docs/book/common_code/architecture.md [hunk 9]
```diff
@@ -31,15 +31,15 @@ watches for create / update / delete events on Resources, and triggers a
...
 state.
 
 The Cluster API consists of a shared set of controllers in order to provide a 
 consistent user experience. In order to support multiple cloud environments, 
-some of these controllers (e.g. `Machine`) call provider specific actuators to 
-do implement the behavior expectived of the controller. Other controllers (e.g.
-`MachineSet`) are generic and operate by interacting with other Cluster API 
-(and Kubernetes) resources.
+some of these controllers (e.g. `Cluster`, `Machine`) call provider specific 
+actuators to implement the the behavior expected of the controller. Other
```
Comment (msohn) [inline docs/book/common_code/architecture.md]: remove duplicate "the"

File: docs/book/common_code/machine_controller.md (hunk 5 above)
Comment (davidewatson) [reply docs/book/common_code/machine_controller.md]: `include-codebook` is more popular (29 vs 2 stars, 22 vs 2 forks) and has seen more recent development (though not by much).

As an example, with this code change:

```
diff --git a/pkg/apis/cluster/v1alpha1/machinedeployment_types.go b/pkg/apis/cluster/v1alpha1/machinedeployment_types.go
index d3017c05d..e27c880da 100644
--- a/pkg/apis/cluster/v1alpha1/machinedeployment_types.go
+++ b/pkg/apis/cluster/v1alpha1/machinedeployment_types.go
@@ -23,6 +23,7 @@ import (
        "sigs.k8s.io/cluster-api/pkg/apis/cluster/common"
 )
 
+/// [marker0]
 // MachineDeploymentSpec defines the desired state of MachineDeployment
 type MachineDeploymentSpec struct {
        // Number of desired machines. Defaults to 1.
@@ -67,6 +68,9 @@ type MachineDeploymentSpec struct {
        ProgressDeadlineSeconds *int32 `json:"progressDeadlineSteeconds,omitempty"`
 }
 
+/// [marker0]
```

The coresponding markdown looks like this:

```
{% method %}
## MachineDeploymentSpec

{% sample lang="go" %}
[import:'marker0'](../../../pkg/apis/cluster/v1alpha1/machinedeployment_types.go)
{% endmethod %}
```

So the resulting markdown is both less readable and more accurate...

File: docs/book/common_code/cluster_controller.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+
+{% method %}
+## ClusterStatus
+
+Some providers use the `APIEndpoint` field to determine when one or more
+masters have been provisioned. This may be necessary before worker nodes
+can be provisioned. For example, `cluster-api-provider-gcp` does [this](
+https://github.com/kubernetes-sigs/cluster-api-provider-gcp/blob/f3145d8810a5c7fc434ddb5577699b4deb1b5fa6/pkg/cloud/google/metadata.go#L43):
+
+```go
```
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/common_code/cluster_controller.md]: I thought so too, however, when I do this it will no longer render. Cf. https://github.com/GitbookIO/theme-api#sample-blocks

File: docs/book/provider_implementations/building_running_and_testing.md [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+ # Run go fmt against code
+ fmt:
+```
+
+## Build and push images
+
+```bash
+dep ensure
+make
+export IMG=quay.io/samsung_cnct/cluster-api-provider-solas
```
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/provider_implementations/building_running_and_testing.md]: I agree that pointing to quay is not ideal, though I did want to use the container name `cluster-api-provider-solas` for consistency with the rest of the document. Would something like this be better?

```
export IMG=<your-docker-registry>/cluster-api-provider-solas
```

Unsure.

File: docs/book/provider_implementations/building_running_and_testing.md (hunk 7 above)
Comment (davidewatson) [reply docs/book/provider_implementations/building_running_and_testing.md]: https://github.com/kubernetes-sigs/cluster-api/issues/567

**Comment by phyber** [comment]
This is looking great. 👍

**Comment by chaosaffe** [comment]
Awesome stuff @davidewatson! Thank you for picking this up

File: docs/book/common_code/cluster_controller.md [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+{% method %}
+## Cluster Actuator Interface
+
+All methods should be idempotent.
+
+`Reconcile()` will be called whenever there is a change to the `Cluster` 
+`Spec`, or after every resync period.
+
+If a `Cluster` resource is deleted, the controller will call the actuator's 
+`Delete()` method until it succeeds, or the finalizer is removed (see below).
```
Comment (chuckha) [inline docs/book/common_code/cluster_controller.md]: Should we link out to https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers on the word finalizer? alternatively we could make sure the CRD page is background reading including advanced topics

File: docs/book/common_code/cluster_controller.md [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+{% panel style="info", title="Important" %}
+To see the actual and current definitions see the [source](#cluster_source).
+{% endpanel %}
+
+{% method %}
+## Cluster
+
+Cluster has 4 fields:
+
+`Spec` contains the desired cluster state specified by the object. While much
```
Comment (chuckha) [inline docs/book/common_code/cluster_controller.md]: This could be a little clearer if the order here matched the order they appeared in the struct below. I think a personal preference is to put the struct first then the explanation after, but I think this ordering works as well

File: docs/book/common_code/machine_controller.md [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+{% panel style="info", title="Important" %}
+To see the actual and current definitions see the [source](#machine_types_source).
+{% endpanel %}
+
+{% method %}
+## Machine
+
+`Machine` has 4 fields:
+
+`Spec` contains the desired cluster state specified by the object. While much
```
Comment (chuckha) [inline docs/book/common_code/machine_controller.md]: same minor feedback as with cluster, ordering of fields to match the struct

File: docs/book/common_code/machine_controller.md [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+		// the stack that provisions it.
+		ready := m.Status.NodeRef != nil || len(m.Annotations) > 0
+		return ready, nil
+```
+
+{% sample lang="go" %}
+```go
+// MachineSpec defines the desired state of Machine
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
```
Comment (chuckha) [inline docs/book/common_code/machine_controller.md]: I opened a cluster-api issue since these have golint issues: https://github.com/kubernetes-sigs/cluster-api/issues/570

File: docs/book/provider_implementations/create_actuators.md [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+// NewActuator creates a new Actuator
+func NewActuator(params ActuatorParams) (*Actuator, error) {
+        return &Actuator{
+                machinesGetter: params.MachinesGetter,
+        }, nil
+}
+
+// Create creates a machine and is invoked by the Machine Controller
+func (a *Actuator) Create(cluster *clusterv1.Cluster, machine *clusterv1.Machine) error {
+        glog.Infof("Creating machine %v for cluster %v.", machine.Name, cluster.Name)
```
Comment (chuckha) [inline docs/book/provider_implementations/create_actuators.md]: Using the standard lib logger instead of glog here would allow for fewer dependencies and fewer 3rd party libraries for new users to learn

File: docs/book/provider_implementations/register_schemes.md [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+                entryLog.Error(err, "unable to set up overall controller manager")
+                os.Exit(1)
+        }
+
+        cs, err := clientset.NewForConfig(cfg)
+        if err != nil {
+                panic(err)
+        }
+
+        clusterActuator, _ := cluster.NewActuator(cluster.ActuatorParams{
```
Comment (chuckha) [inline docs/book/provider_implementations/register_schemes.md]: this should panic like it does on line 65, same with machine actuator

**Comment by randomvariable** [comment]
This looks pretty great.

For remaining tasks, we can track these in an "UMBRELLA: Gitbook" issue.

Just as a test, was able to get this to push to [Github pages](https://randomvariable.github.io/cluster-api/). Would require [renaming the current docs](https://github.com/randomvariable/cluster-api/commit/cc81159a90fd675851ee7c9c58f0c3220ed37e37) directory to work.

File: docs/book/common_code/cluster_controller.md (hunk 13 above)
Comment (davidewatson) [reply docs/book/common_code/cluster_controller.md]: Thinking more about this, I am not sure I agree. By placing the `Spec` and `Status` descriptions first aren't we are emphasizing the portion of the API which matters most to the implementor?

Comment (chuckha) [reply docs/book/common_code/cluster_controller.md]: i can see it both ways, totally fine to keep it as is

File: docs/book/common_code/cluster_controller.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/common_code/cluster_controller.md]: The reason this is not a sample is we want the text to appear in the left column as part of the description, and not in the right as an example...

File: docs/book/common_code/machineset_controller.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/common_code/machineset_controller.md]: Fixed with my latest commit.

File: docs/book/common_code/machineset_controller.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/common_code/machineset_controller.md]: Fixed with my latest commit.

File: docs/book/common_code/cluster_controller.md [hunk 18]
```diff
@@ -0,0 +1,156 @@
...
+`Delete()` method until it succeeds, or the finalizer is removed (see below).
+
+**TODO**: Determine what the current resync period is.
+
+{% sample lang="go" %}
+```go
+// Actuator controls clusters on a specific infrastructure. All
+// methods should be idempotent unless otherwise specified.
+type Actuator interface {
+	// Create or update the cluster
```
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/common_code/cluster_controller.md]: Fixed with my latest commit.

**Comment by davidewatson** [comment]
@randomvariable: I like the features of firebase, and see some value in maintaining consistency with kubebuilder, but gh-pages are free and therefore more likely to be replicated by other kubebuilder projects. Maybe we can discuss briefly tomorrow and make a decision. I'll also create an umbrella issue for the major items which are missing. Thanks!

**Comment by davidewatson** [comment]
I have updated the deployment instructions. Once this is merged someone with admin privileges will need to:

 - Create the gh-pages branch:

```
git subtree push --prefix docs/book/_book origin gh-pages
```

- Enable gh-pages from the Setting pane:

github pages configuration

/assign @roberthbailey

**Comment by roberthbailey** [comment]
I'm happy to approve once someone lgtm's this who has reviewed it somewhat carefully (I took a quick pass on the docs but from the call today it sounded like other folks had done a much better job reviewing them than I have).

File: docs/book/_book/.firebase/hosting.cHVibGlj.cache [hunk 19]
```diff
@@ -0,0 +1,56 @@
+Dockerfile,1540782337297,8d44c51c2eee089aba3bea9acc2bbfa777930221543a461ba70e8f91faaa4476
```
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/_book/.firebase/hosting.cHVibGlj.cache]: Good catch!

File: docs/book/_book/GLOSSARY.html [hunk 20]
```diff
@@ -0,0 +1,468 @@
...
+<html lang="" >
+    <head>
+        <meta charset="UTF-8">
+        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
+        <title> · The Cluster API Book</title>
+        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+        <meta name="description" content="">
+        <meta name="generator" content="GitBook 3.2.3">
+        <meta name="author" content="SIG Lifecycle Team">
+
```
[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/_book/GLOSSARY.html]: It is strange. Maybe there is a stray new line (or two!) somewhere. I'll look into it.

[the remark this answers is not in the document]
Comment (msohn) [reply docs/book/_book/GLOSSARY.html]: Are you aware of this discussion on gitbook's future as an open source project:
https://github.com/GitbookIO/gitbook/issues/1808
https://github.com/GitbookIO/gitbook/issues/2148

it seems the new version does not support self-hosting anymore
https://docs.gitbook.com/v2-changes/important-differences

[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/_book/GLOSSARY.html]: It may be we that we have to or should migrate to netlify in the future. @timothysc suggested this last week since it is what k8s uses.

If we do it might also be nice to migrate kubebuilder as well. I am unsure if there is interest on that front but I think there is value in leveraging kubebuilder to the extent possible since this will make it easier for others to create high quality CRDs. @pwittrock 

For now I'd like to get this merged so that we can parallize and iterate. @detiber, @roberthbailey, thoughts?

[the remark this answers is not in the document]
Comment (detiber) [reply docs/book/_book/GLOSSARY.html]: I see no reason to block this now, but it'll definitely be worth tracking it for later followup.

[the remark this answers is not in the document]
Comment (msohn) [reply docs/book/_book/GLOSSARY.html]: +1 for merging, I didn't intend to block this, content is more important than the tool

k8s documentation is written in markdown, processed using Hugo and served using netlify
https://kubernetes.io/docs/contribute/start/#the-basics-about-our-docs

[the remark this answers is not in the document]
Comment (davidewatson) [reply docs/book/_book/GLOSSARY.html]: Thanks for the link!

**Comment by roberthbailey** [comment]
Let's get this in and we can iterate from there.

/approve

**Comment by roberthbailey** [comment]
I tried to follow the instructions [above](https://github.com/kubernetes-sigs/cluster-api/pull/566#issuecomment-434830900) to push to a branch to serve on gh pages but it didn't work:

```
$ git subtree push --prefix docs/book/_book upstream gh-pages
git: 'subtree' is not a git command. See 'git --help'.
```

Thoughts?

**Comment by msohn** [comment]
Usually git subtree comes in the ```./share/git-core/contrib/``` folder of a git installation.
On Mac it just works for me using git 2.19.1:

```
$ git subtree
usage: git subtree add   --prefix=<prefix> <commit>
   or: git subtree add   --prefix=<prefix> <repository> <ref>
   or: git subtree merge --prefix=<prefix> <commit>
   or: git subtree pull  --prefix=<prefix> <repository> <ref>
   or: git subtree push  --prefix=<prefix> <repository> <ref>
   or: git subtree split --prefix=<prefix> <commit...>

    -h, --help            show the help
    -q                    quiet
    -d                    show debug messages
    -P, --prefix ...      the name of the subdir to split out
    -m, --message ...     use the given message as the commit message for the merge commit

options for 'split'
    --annotate ...        add a prefix to commit message of new commits
    -b, --branch ...      create a new branch from the split subtree
    --ignore-joins        ignore prior --rejoin commits
    --onto ...            try connecting new tree to an existing one
    --rejoin              merge the new branch back into HEAD

options for 'add', 'merge', and 'pull'
    --squash              merge subtree changes as a single commit
```

**Comment by davidewatson** [comment]
My guess is either the`git-subtree` command is not installed or it is not in your `$PATH` (see https://git-scm.com/docs/git#git---list-cmdsgroupgroup82308203). On MacOS, when `git-subtree` is installed using brew, I find it here:

```
rodin:~ dwat$ which git-subtree
/usr/local/bin/git-subtree
rodin:~ dwat$ 
```

The instructions I found for installing `git-subtree` on [StackOverflow](https://stackoverflow.com/questions/17797328/why-is-the-git-subtree-not-working-even-though-i-have-git-1-8-3-2-installed-on) say to download git and compile from source if it is not available as a package for your distro. You can also install only `git-subtree` using these [instructions](https://github.com/git/git/blob/53f9a3e157dbbc901a02ac2c73346d375e24978c/contrib/subtree/INSTALL).

**Comment by roberthbailey** [comment]
@davidewatson - I gave you write access to the repo; can you run the git subtree commands and get the branch pushed? I have a google-installed version of git and don't feel like breaking it right now to get the subtree command working via brew. :)
